plugins/system/cookiemanager/cookiemanager.php

```diff
@@ -0,0 +1,342 @@
+
+			foreach ($cookies as $key => $value)
+			{
+				if (!empty($value))
+				{
+					if ($catValue->id == $value->id)
+					{
+						if ($value->exp_period == -1)
+						{
+							$value->exp_period = "Forever";
+							$value->exp_value = "";
+						}
+						elseif ($value->exp_period == 0)
+						{
+							$value->exp_period = "Session";
+							$value->exp_value = "";
+						}
+
+						$table .= '<tr>'
+						. '<td>' . $value->cookie_name . '</td>'
+						. '<td>' . $value->cookie_desc . '</td>'
+						. '<td>' . $value->exp_value . ' ' . $value->exp_period . '</td>'
+						. '</tr>';
+					}
+				}
+			}
+
+			$table .= '</tbody></table>';
+			$prefBody .= $table . '</div>';
+		}
+
+		$this->preferences = HTMLHelper::_(
+			'bootstrap.renderModal',
+			'preferences',
+			[
+				'title' => Text::_('COM_COOKIEMANAGER_PREFERENCES_TITLE'),
+				'footer' => '<button type="button" id="prefConfirmChoice" class="btn btn-info" data-bs-dismiss="modal">'
+					. Text::_('COM_COOKIEMANAGER_CONFIRM_CHOICE_BUTTON_TEXT') . '</button>'
+					. '<button type="button" class="btn btn-info" data-bs-dismiss="modal">'
+					. Text::_('COM_COOKIEMANAGER_ACCEPT_BUTTON_TEXT') . '</button>'
+			],
+			$prefBody
+		);
+
+		$query = $db->getQuery(true)
+			->select($db->quoteName(['a.type','a.position','a.code','a.catid']))
+			->from($db->quoteName('#__categories', 'c'))
+			->join(
+				'RIGHT',
+				$db->quoteName('#__cookiemanager_scripts', 'a') . ' ON ' . $db->quoteName('c.id') . ' = ' . $db->quoteName('a.catid') . 'AND' . $db->quoteName('a.published') . ' =  1'
+			);
+
+			$db->setQuery($query);
+			$this->scripts = $db->loadObjectList();
+
+		foreach ($this->category as $catKey => $catValue)
+		{
+			if (!isset($_COOKIE['cookie_category_' . $catValue->alias]) || $_COOKIE['cookie_category_' . $catValue->alias] === 'false')
+			{
+				$this->script[$catValue->alias] = [];
+
+				foreach ($this->scripts as $key => $value)
+				{
+					if ($catValue->id == $value->catid)
+					{
+						array_push($this->script[$catValue->alias], $value);
+					}
+				}
+			}
+		}
+
+				$this->app->getDocument()->addScriptOptions('code', $this->script);
+
+	}
+
+	/**
+	 * Echo the modal and button.
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function onAfterRespond()
+	{
+		if (!$this->app->isClient('site'))
+		{
+			return;
+		}
+
+		echo $this->cookieBanner;
+		echo $this->preferences;
+		echo '<button class="preview btn btn-info" data-bs-toggle="modal" data-bs-target="#cookieBanner">' . Text::_('COM_COOKIEMANAGER_PREVIEW_BUTTON_TEXT') . '</button>';
+
+		foreach ($this->category as $catKey => $catValue)
+		{
+			if (isset($_COOKIE['cookie_category_' . $catValue->alias]) && $_COOKIE['cookie_category_' . $catValue->alias] === 'true')
+			{
+				$this->script[$catValue->alias] = [];
+
+				foreach ($this->scripts as $key => $value)

```

Comment:
This foreach looks a bit "smelly", I think this has a potential to break...we have to check.